src/main.rs

```diff
@@ -1,124 +1,96 @@
-        None => root_dir.join("config.toml"),
-    };
+    let cli = Cli::parse();
+    let root_dir = cli
+        .root
+        .canonicalize()
+        .unwrap_or_else(|_| panic!("Cannot find root directory: {}", cli.root.display()));
+    let config_file = cli
+        .config
+        .map(|path| {
+            path.canonicalize()
+                .unwrap_or_else(|_| panic!("Cannot find config file: {}", path.display()))
+        })
+        .unwrap_or_else(|| root_dir.join("config.toml"));
 
-    match matches.subcommand() {
-        ("init", Some(matches)) => {
-            let force = matches.is_present("force");
-            match cmd::create_new_project(matches.value_of("name").unwrap(), force) {
-                Ok(()) => (),
-                Err(e) => {
-                    console::unravel_errors("Failed to create the project", &e);
-                    ::std::process::exit(1);
-                }
-            };
-        }
-        ("build", Some(matches)) => {
+    match cli.command {
+        Command::Init { name, force } => match cmd::create_new_project(&name, force) {
+            Ok(()) => (),
+            Err(e) => {
+                console::unravel_errors("Failed to create the proejct", &e);
+                std::process::exit(1);
+            }
+        },
+        Command::Build { base_url, output_dir, drafts } => {
             console::info("Building site...");
             let start = Instant::now();
-            let output_dir = matches.value_of("output_dir").map(|output_dir| Path::new(output_dir));
             match cmd::build(
                 &root_dir,
                 &config_file,
-                matches.value_of("base_url"),
-                output_dir,
-                matches.is_present("drafts"),
+                base_url.as_deref(),
+                output_dir.as_deref(),
+                drafts,
             ) {
                 Ok(()) => console::report_elapsed_time(start),
                 Err(e) => {
                     console::unravel_errors("Failed to build the site", &e);
-                    ::std::process::exit(1);
+                    std::process::exit(1);
                 }
-            };
+            }
         }
-        ("serve", Some(matches)) => {
-            let interface = matches.value_of("interface").unwrap_or("127.0.0.1");
-            let mut port: u16 = match matches.value_of("port").unwrap_or("1111").parse() {
-                Ok(x) => x,
-                Err(_) => {
-                    console::error("The request port needs to be an integer");
-                    ::std::process::exit(1);
-                }
-            };
-            let open = matches.is_present("open");
-            let include_drafts = matches.is_present("drafts");
-            let fast = matches.is_present("fast");
-
-            // Default one
+        Command::Serve { interface, mut port, output_dir, base_url, drafts, open, fast } => {
```

Comment:
what error message do users get when the port is not a valid number for example with clap3?

Reply:
by default whatever the error from the `FromStr` impl for the type (here `u16`) displays, e.g.:

![image](https://user-images.githubusercontent.com/5671049/148046408-f2e82890-56b5-4317-b341-0dd9df6afc6d.png)